src/metabolic_ninja/worker/tasks.py

```diff
@@ -0,0 +1,200 @@
+# Copyright (c) 2018, Novo Nordisk Foundation Center for Biosustainability,
+# Technical University of Denmark.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import json
+import logging
+
+import cameo.api
+from cobra.io.dict import metabolite_to_dict, reaction_to_dict
+from sendgrid import SendGridAPIClient
+from sendgrid.helpers.mail import Email, Mail, Personalization
+
+from . import designer
+from .data import Job
+from .decorators import TaskFailedException, task
+
+
+logger = logging.getLogger(__name__)
+
+
+def design(connection, channel, delivery_tag, body, ack_message):
+    """Run the metabolic ninja design workflow."""
+    job = Job.deserialize(json.loads(body))
+
+    try:
+        job.save(status="STARTED")
+
+        logger.info(f"Initiating new design workflow")
+
+        logger.debug("Starting task: Find product")
+        product = find_product(job)
+        logger.debug(f"Task finished; found product: {product}")
+
+        logger.debug("Starting task: Find pathways")
+        pathways = find_pathways(job, product)
+        logger.debug(
+            f"Task finished: Find pathways, found {len(pathways)} pathways"
+        )
+
+        logger.debug("Starting task: Optimize")
+        results = optimize(job, pathways)
+        logger.debug("Task finished: Optimize")
+
+        # Save the results
+        job.save(status="SUCCESS", result=results)
+
+        _notify(job)
+    except TaskFailedException:
+        # Exceptions are handled in the child processes, so there's nothing to do here.
+        # Just abort the workflow and get ready for new jobs.
+        logger.info(
+            f"Task failed; aborting workflow and restarting consumption from queue"
+        )
+    finally:
+        # Acknowledge the message, whether it failed or not.
+        connection.add_callback_threadsafe(
+            functools.partial(ack_message, channel, delivery_tag)
+        )
+
+
+@task
+def find_product(job):
+    # Find the product name via the cameo designer. In a future far, far away
+    # this should be a call to a web service.
+    # TODO: update db state
+    return cameo.api.design.translate_product_to_universal_reactions_model_metabolite(
+        job.product_name, job.source
+    )
+
+
+@task
+def find_pathways(job, product):
+    # TODO: update db state
```

Comment:
Is this TODO still relevant?

Reply:
It's related to the idea of updating progress for an ongoing job. We're tracking that in zenhub now though so the comment in code is not really necessary anymore, I'll remove it.